lib/http2.c

```diff
@@ -1166,6 +1166,9 @@ static ssize_t data_source_read_callback(nghttp2_session *session,
 
   nread = CURLMIN(stream->upload_len, length);
   if(nread > 0) {
+    if(length < 2048 &&
+       (stream->upload_len >= 2048 || stream->upload_left >= 2048))
+      return NGHTTP2_ERR_DEFERRED;
```

Comment:
This obviously does not work if server allows less than 2048 of window.  But in practice, it is quite rare.

In some particular scenario, curl goes into busy loop.  For example, initial window size is 65535 + 2048, and if server never send window update, curl goes into busy loop.  Server that never send window update is kinda broken though.

Reply:
I've removed the commit from this PR.

Is there a way to return NGHTTP2_ERR_DEFERRED or something else (PAUSE?) to avoid spinning and instead wait for an external event to re-wake/retry the stream, i.e. a WINDOW_UPDATE?

Reply:
NGHTTP2_ERR_DEFERRED actually removes the stream from the tx queue.  But curl calls nghttp2_session_resume_stream when data is available, which brings back the stream into the queue.

Reply:
If I understand you correctly, you are suggesting that the places in curl `lib/http2.c` which call `nghttp2_session_resume_data()` could, if desired, check for window sizes above a certain threshold (low watermark) instead of the current behavior which checks for window sizes > 0.

Reply:
No, I am not suggesting that.  I just answered you question about the interaction between NGHTTP2_ERR_DEFERRED and nghttp2_session_resume_stream.

Reply:
In https://github.com/nghttp2/nghttp2/issues/1722, you said this was an application's problem, not the nghttp2 library.  Is there any doc I could read about the recommended way to use libnghttp2 and have the application manage the HTTP/2 window sizes?  lighttpd uses an HTTP/2 implementation that I wrote; lighttpd does not use libnghttp2.  However, curl does use libnghttp2 and as I am not an expert on libnghttp2, and since libnghttp2 is not going to help applications avoid degenerative behaviors like the issue reported here, I am asking for guidance/examples on how an application such as curl should interact with libnghttp2 to manage window sizes.